core/services/vrf/v2/listener_v2_log_processor.go

```diff
@@ -0,0 +1,1237 @@
+)
+
+// Returns all the confirmed logs from the provided pending queue by subscription
+func (lsn *listenerV2) getConfirmedLogsBySub(latestHead uint64, pendingRequests []pendingRequest) map[string][]pendingRequest {
+	vrfcommon.UpdateQueueSize(lsn.job.Name.ValueOrZero(), lsn.job.ExternalJobID, lsn.coordinator.Version(), uniqueReqs(pendingRequests))
+	var toProcess = make(map[string][]pendingRequest)
+	for _, request := range pendingRequests {
+		if lsn.ready(request, latestHead) {
+			toProcess[request.req.SubID().String()] = append(toProcess[request.req.SubID().String()], request)
+		}
+	}
+	return toProcess
+}
+
+func (lsn *listenerV2) ready(req pendingRequest, latestHead uint64) bool {
+	// Request is not eligible for fulfillment yet
+	if req.confirmedAtBlock > latestHead {
+		return false
+	}
+
+	if lsn.job.VRFSpec.BackoffInitialDelay == 0 || req.attempts == 0 {
+		// Backoff is disabled, or this is the first try
+		return true
+	}
+
+	return time.Now().UTC().After(
+		nextTry(
+			req.attempts,
+			lsn.job.VRFSpec.BackoffInitialDelay,
+			lsn.job.VRFSpec.BackoffMaxDelay,
+			req.lastTry))
+}
+
+func nextTry(retries int, initial, max time.Duration, last time.Time) time.Time {
+	expBackoffFactor := math.Pow(backoffFactor, float64(retries-1))
+
+	var delay time.Duration
+	if expBackoffFactor > float64(max/initial) {
+		delay = max
+	} else {
+		delay = time.Duration(float64(initial) * expBackoffFactor)
+	}
+	return last.Add(delay)
+}
+
+// Remove all entries 10000 blocks or older
+// to avoid a memory leak.
+func (lsn *listenerV2) pruneConfirmedRequestCounts() {
+	min := lsn.blockNumberToReqID.FindMin()
+	for min != nil {
+		m := min.(fulfilledReqV2)
+		if m.blockNumber > (lsn.getLatestHead() - 10000) {
+			break
+		}
+		delete(lsn.respCount, m.reqID)
+		lsn.blockNumberToReqID.DeleteMin()
+		min = lsn.blockNumberToReqID.FindMin()
+	}
+}
+
+// Determine a set of logs that are confirmed
+// and the subscription has sufficient balance to fulfill,
+// given a eth call with the max gas price.
+// Note we have to consider the pending reqs already in the txm as already "spent" link or native,
+// using a max link or max native consumed in their metadata.
+// A user will need a minBalance capable of fulfilling a single req at the max gas price or nothing will happen.
+// This is acceptable as users can choose different keyhashes which have different max gas prices.
+// Other variables which can change the bill amount between our eth call simulation and tx execution:
+// - Link/eth price fluctuation
+// - Falling back to BHS
+// However the likelihood is vanishingly small as
+// 1) the window between simulation and tx execution is tiny.
+// 2) the max gas price provides a very large buffer most of the time.
+// Its easier to optimistically assume it will go though and in the rare case of a reversion
+// we simply retry TODO: follow up where if we see a fulfillment revert, return log to the queue.
```

Comment:
Should we consider this:
> follow up where if we see a fulfillment revert, return log to the queue "up to a certain number of retries"

Reply:
That's been a TODO for a while, I won't address it in this PR.